mily/runengine.py

```diff
@@ -0,0 +1,305 @@
+        # Do not allow plans to be set while RunEngine is active
+        if self.engine and self.engine.state != 'idle':
+            logger.exception("Can not change the configured plan while the "
+                             "RunEngine is running!")
+            return
+        # Store our plan internally
+        self._plan = plan
+        # Register a new call command
+        self.command_registry['Start'] = (lambda engine: engine(self.plan()))
+
+    @property
+    def engine(self):
+        """
+        Underlying RunEngine object
+        """
+        return self._engine
+
+    @engine.setter
+    def engine(self, engine):
+        logger.debug("Storing a new RunEngine object")
+        # Do not allow engine to be swapped while RunEngine is active
+        if self._engine and self._engine.state != 'idle':
+            raise RuntimeError("Can not change the RunEngine while the "
+                               "RunEngine is running!")
+        # Create a kicker, not worried about doing this multiple times as this
+        # is checked by `install_qt_kicker` itself
+        install_qt_kicker(update_rate=self.update_rate)
+        engine.state_hook = self.on_state_change
+        # Connect signals
+        self._engine = engine
+        self.engine_state_change.connect(self.label.on_state_change)
+        self.engine_state_change.connect(self.control.on_state_change)
+        self.control.clicked.connect(self.command)
+        # Run callbacks manually to initialize widgets. We can not emit the
+        # signal specifically because we can not emit signals in __init__
+        state = self._engine.state
+        self.label.on_state_change(state, None)
+        self.control.on_state_change(state, None)
+
+    def on_state_change(self, state, old_state):
+        """
+        Report a state change of the RunEngine
+
+        This is added directly to the `RunEngine.state_hook` and emits the
+        `engine_state_change` signal.
+
+        Parameters
+        ----------
+        state: str
+
+        old_state: str
+
+        """
+        self.engine_state_change.emit(state, old_state)
+
+    def command(self, command_button):
+        """
+        Accepts commands and instructs the RunEngine accordingly
+
+        Parameters
+        ----------
+        command : str
+            Name of the command in the :attr:`.command_registry:`
+        """
+        command = command_button.text()
+
+        logger.info("Requested command %s for RunEngine", command)
+        # Load thefunction from registry
+        try:
+            func = self.command_registry[command]
+        except KeyError as exc:
+            logger.exception('Unrecognized command for RunEngine -> %s',
+                             exc)
+            return
+        # Execute our loaded function
+        try:
+            func(self.engine)
+        # Pausing raises an exception
+        except RunEngineInterrupted as exc:
```

Comment:
Might be worth catching generic `Exception` objects here. Nobody likes a GUI crash and this is where any bubbles from the execution of the plan will bubble up. There might also need to be some logic to make sure it is ready to take another plan afterwards (teardown and re-up of a failed RunEngine execution)? 🤔